Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java

```diff
@@ -525,3 +577,117 @@ public String verbatimString() {
         return string;
     }
 }
+
+final class ConcatDafnySequence<T> extends DafnySequence<T> {
+    // INVARIANT: Either these are both non-null and ans is null or both are
+    // null and ans is non-null.
+    private DafnySequence<T> left, right;
+    // XXX We're storing the combined sequence as a DafnySequence, not an array,
+    // because it might need a backing array of primitive type.  Unfortunately,
+    // this means the resulting sequence is doomed to have a bit of overhead (an
+    // extra indirect function call) on every call to select().  Hard to see how
+    // to fix this without requiring that *every* DafnySequence suffer some
+    // overhead.
+    private DafnySequence<T> ans = null;
+    private final int length;
+
+    ConcatDafnySequence(DafnySequence<T> left, DafnySequence<T> right) {
+        this.left = left;
+        this.right = right;
+        this.length = left.length() + right.length();
+    }
+
+    private DafnySequence<T> resolve() {
+        if (ans == null) {
+            ans = computeElements();
+            // Allow left and right to be garbage-collected
+            left = null;
+            right = null;
+        }
+
+        return ans;
+    }
+
+    @Override
+    public T select(int i) {
+        return resolve().select(i);
+    }
+
+    @Override
+    public int length() {
+        return length;
+    }
+
+    @Override
+    public DafnySequence<T> update(int i, T t) {
+        return resolve().update(i, t);
+    }
+
+    @Override
+    public DafnySequence<T> subsequence(int lo, int hi) {
+        return resolve().subsequence(lo, hi);
+    }
+
+    @Override
+    Copier<T> newCopier(int length) {
+        return resolve().newCopier(length);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return resolve().iterator();
+    }
+
+    @Override
+    public int hashCode() {
+        return resolve().hashCode();
+    }
+
+    private DafnySequence<T> computeElements() {
+        // Somewhat arbitrarily, the copier will be created by the leftmost
+        // sequence.  TODO: Have enough run-time type information that this is
```

Comment:
An `ArrayDafnySequence` is just a wrapper around `Object[]` so all the contained elements are boxed.  What's an example of a disastrous combination?

Reply:
The disaster would be an  `ArrayDafnySequence<Character>` being appended to, say, a large `StringDafnySequence`, which would result in a large `ArrayDafnySequence<Character>` and a whole lot of boxing.

Reply:
You mean the `StringDafnySequence` is appended to the `ArrayDafnySequence<Character>` right?  That makes sense.